src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -395,27 +400,45 @@ public class ForkJoinPool extends AbstractExecutorService {
-     * bits. It is updated only via bitwise atomics (getAndBitwiseOr).
+     * subfields.
+     *
+     * Field "runState" and per-WorkQueue field "phase" play similar
+     * roles, as lockable, versioned counters. Field runState also
+     * includes monotonic event bits (SHUTDOWN, STOP, and TERMINATED).
+     * The version tags enable detection of state changes (by
+     * comparing two reads) modulo bit wraparound. The bit range in
+     * each case suffices for purposes of determining quiescence,
+     * termination, avoiding ABA-like errors, and signal control, most
+     * of which are ultimately based on at most 15bit ranges (due to
+     * 32767 max total workers). RunState updates do not need to be
+     * atomic with respect to ctl updates, but because they are not,
+     * some care is required to avoid stalls. The seqLock properties
+     * detect changes and conditionally upgrade to coordinate with
+     * updates. It is typically held for less than a dozen
+     * instructions unless the queue array is being resized, during
+     * which contention is rare. To be conservative, lockRunState is
+     * implemented as a spin/sleep loop. Here and elsewhere spin
+     * constants are short enough to apply even on systems with few
+     * available processors.  In addition to checking pool status,
+     * reads of runState sometimes serve as acquire fences before
+     * reading other fields.
+     *
+     * Field "parallelism" holds the target parallelism (normally
+     * corresponding to pool size). Users can dynamically reset target
+     * parallelism, but is only accessed when signalling or awaiting
+     * work, so only slowly has an effect in creating threads or
+     * letting them time out and terminate when idle.
      *
      * Array "queues" holds references to WorkQueues.  It is updated
-     * (only during worker creation and termination) under the
-     * registrationLock, but is otherwise concurrently readable (often
-     * prefaced by a volatile read of mode to check termination, that
-     * is required anyway, and serves as an acquire fence). To
+     * (only during worker creation and termination) under the runState
+     * lock. It is otherwise concurrently readable but reads for use
+     * in scans (see below) are always prefaced by a volatile read of
+     * runState (or equivalent constructions), ensuring that its state is
+     * current at the point it is used (which is all we require). To
      * simplify index-based operations, the array size is always a
      * power of two, and all readers must tolerate null slots.  Worker
-     * queues are at odd indices. Worker ids masked with SMASK match
-     * their index. Shared (submission) queues are at even
+     * queues are at odd indices. Worker phase ids masked with SMASK
+     * match their index. Shared (submission) queues are at even
      * indices. Grouping them together in this way simplifies and
```

Comment:
More-or-less. At top-level, both kinds of queues should be sampled with approx the same probability, which is simpler if they are all in the same array. But we also need to identify what kind they are without looking at them, leading to odd/even scheme. One disadvantage is that there are usually many fewer submission queues, so there can be a lot of wasted probes (null slots). But still cheaper than alternatives because of cache locality etc.

Reply:
(I also added some similar wording to the internal docs.)